service/history/replicationTaskFetcher.go

```diff
@@ -165,15 +152,64 @@ func (f *ReplicationTaskFetchersImpl) Stop() {
 		return
 	}
 
+	currentCluster := f.clusterMetadata.GetCurrentClusterName()
+	f.clusterMetadata.UnRegisterMetadataChangeCallback(currentCluster)
+	f.fetchersLock.Lock()
+	defer f.fetchersLock.Unlock()
 	for _, fetcher := range f.fetchers {
 		fetcher.Stop()
 	}
 	f.logger.Info("Replication task fetchers stopped.")
 }
 
-// GetFetchers returns all the fetchers
-func (f *ReplicationTaskFetchersImpl) GetFetchers() []ReplicationTaskFetcher {
-	return f.fetchers
+func (f *ReplicationTaskFetchersImpl) GetOrCreateFetcher(clusterName string) ReplicationTaskFetcher {
+	f.fetchersLock.Lock()
+	defer f.fetchersLock.Unlock()
+
+	fetcher, ok := f.fetchers[clusterName]
+	if ok {
+		return fetcher
+	}
+	return f.createReplicationFetcherLocked(clusterName)
+}
+
+func (f *ReplicationTaskFetchersImpl) createReplicationFetcherLocked(clusterName string) ReplicationTaskFetcher {
+	currentCluster := f.clusterMetadata.GetCurrentClusterName()
+	remoteAdminClient := f.clientBean.GetRemoteAdminClient(clusterName)
+	fetcher := newReplicationTaskFetcher(
+		f.logger,
+		clusterName,
+		currentCluster,
+		f.config,
+		remoteAdminClient,
+	)
+	fetcher.Start()
+	f.fetchers[clusterName] = fetcher
+	return fetcher
+}
+
+func (f *ReplicationTaskFetchersImpl) listenClusterMetadataChange() {
+	currentCluster := f.clusterMetadata.GetCurrentClusterName()
+	f.clusterMetadata.RegisterMetadataChangeCallback(
+		currentCluster,
+		func(oldClusterMetadata map[string]*cluster.ClusterInformation, newClusterMetadata map[string]*cluster.ClusterInformation) {
+			f.fetchersLock.Lock()
+			defer f.fetchersLock.Unlock()
+
+			for clusterName := range newClusterMetadata {
+				if clusterName == currentCluster {
+					continue
+				}
+				if fetcher, ok := f.fetchers[clusterName]; ok {
+					fetcher.Stop()
+					delete(f.fetchers, clusterName)
+				}
```

Comment:
same problem here, should delete existing one if they no longer exists in newClusterMetadata

Reply:
Same